### PR TITLE
[POC] Use a hooks interface to support executing custom logic after client operations

### PR DESCRIPTION
**Pull request by chatton**
## Description

This PR enables hooks to allow for arbitrary code execution upon client events (upgrade/create/update)

ref: https://github.com/cosmos/ibc-go/issues/2880

### Commit Message / Changelog Entry

```bash
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/02-client/types/exp/hooks.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+)
+
+// Hooks specifies methods to enable writing custom logic which executes
+// after specific client lifecycle methods.
+// Note: this interface is experimental and could be removed in favour of a different solution, requiring code changes.
+type Hooks interface {
+	// OnClientCreated is executed upon client creation.
+	OnClientCreated(ctx sdk.Context, clientID string) error
+	// OnClientUpdated is executed when a client is updated.
+	OnClientUpdated(ctx sdk.Context, clientID string, consensusHeights []exported.Height) error
```
Comment (AdityaSripal) [inline modules/core/02-client/types/exp/hooks.go:16]:
```suggestion
	OnClientUpdated(ctx sdk.Context, clientID string, clientMsg exported.ClientMessage, consensusHeights []exported.Height) error
```

File: modules/core/02-client/types/exp/hooks.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	// OnClientUpgraded is executed when a client is upgraded.
+	OnClientUpgraded(ctx sdk.Context, clientID string) error
+}
```
Comment (AdityaSripal) [inline modules/core/02-client/types/exp/hooks.go:19]: I think a hook for Misbehaviour is also necessary. CCV has expressed interest in something like this

`OnClientMisbehaviour(ctx sdk.Context, clientID string, clientMsg exported.ClientMessage)`

**Review by AdityaSripal — COMMENTED** [review]
Very in favor of this direction!!!

File: modules/core/02-client/types/exp/hooks.go (hunk 1 above)
Comment (AdityaSripal) [reply modules/core/02-client/types/exp/hooks.go:16]: This is so the hook also has access to the header which in the tendermint case has the unhashed validator set as well as other potentially useful information like `DataHash`

File: modules/core/02-client/types/exp/hooks.go
Comment (AdityaSripal) [inline modules/core/02-client/types/exp/hooks.go:1]: I think `exported` is a better name than `exp`

Comment (chatton) [reply modules/core/02-client/types/exp/hooks.go:1]: this was `experimental` rather than `exported`

Comment (AdityaSripal) [reply modules/core/02-client/types/exp/hooks.go:1]: Is the goal to eventually have a release where this is under the experimental directory?

Comment (chatton) [reply modules/core/02-client/types/exp/hooks.go:1]: We can name the directory to be more explicit. Although this PR as it currently exists is not sufficient to achieve what is being achieved by using an `expected keepers` interface. Hooks functionality ensures core ibc logic executes, but in the interface implementation, that logic is being changed i.e. clients on *not* frozen when misbehaviour is detected. We still need to figure out the best way of enabling this.

Comment (AdityaSripal) [reply modules/core/02-client/types/exp/hooks.go:1]: Theoretically, if the hook has access to the client keeper; they can simply unfreeze the client again. Though perhaps that's a bit hacky

Comment (chatton) [reply modules/core/02-client/types/exp/hooks.go:1]: [this](https://github.com/cosmos/ibc-go/pull/2967#issuecomment-1458949771) comment gives more details about possible issues.

**Comment by crodriguezvega** [comment]
Thanks for drafting this alternative solution, @chatton. Closing the PR for now for the same reason as [here](https://github.com/cosmos/ibc-go/pull/2967#issuecomment-1582036087).

**Pull request closed without merging**